pkg/activator/certificate/cache.go

```diff
@@ -40,23 +41,64 @@ import (
 type CertCache struct {
 	secretInformer v1.SecretInformer
 	logger         *zap.SugaredLogger
+	trustConfig    netcfg.Trust
 
-	certificate *tls.Certificate
-	TLSConf     tls.Config
+	certificate   *tls.Certificate
+	ClientTLSConf tls.Config
+	ServerTLSConf tls.Config
 
 	certificatesMux sync.RWMutex
+	caCerts         []*x509.Certificate
+}
+
+func (cr *CertCache) init() {
+	cr.ClientTLSConf.ServerName = certificates.LegacyFakeDnsName
+	cr.ClientTLSConf.MinVersion = tls.VersionTLS13
+	cr.ClientTLSConf.RootCAs = x509.NewCertPool()
```

Comment:
CertPools aren't thread safe - you might instead want to generate a config per connection using `tls.Config.GetConfigForClient`

Reply:
Will using GetConfigForClient help us here?

GetConfigForClient can be used on the server side only. 
We could use it to place a mutex that will ensure that during GetConfigForClients we will:
- Lock
- Clone the cert pool
- Unlock
Is this what you have in mind?

We will not have protection for cr.ClientTLSConf.RootCAs though. 

Would we not also be ok without locking the pool?
Is it wrong to assume that the worst that can happen is for a connection to be rejected in a rare case where it was started while we were updating the pool with a new CA (in which case the behavior is undefined and it is possible that the pool will not include the CA reported by the other side)?

Reply:
> Will using GetConfigForClient help us here?

Yup - it will allow us to return different tls.Config for a new connection. This will help with picking up new CA cert changes

> We could use it to place a mutex that will ensure that during GetConfigForClients we will: ...


> We will not have protection for cr.ClientTLSConf.RootCAs though.

We have control which tls.Config is used in an http.Client when making outbound requests. If the CA changes we need to recreate these clients to use these new certs

> Would we not also be ok without locking the pool?

panics can occur with concurrent reads and writes - you need to see if the certpool allows certain methods can be called by multiple readers and writers. I'm assuming it's not because usually it's mentioned in godoc which I don't see.

Reply:
> >  We will not have protection for cr.ClientTLSConf.RootCAs though.

> We have control which tls.Config is used in an http.Client when making outbound requests. If the CA changes we need to recreate these clients to use these new certs

nice, we can use this.
It does require that we move the lock such that it is available at the handler as well - I will move the lock to transport where the dialTLSContext is used.